File: 33.search-in-rotated-sorted-array.cpp
/*
 * @lc app=leetcode id=33 lang=cpp
 *
 * [33] Search in Rotated Sorted Array
 */
#include <bits/stdc++.h>

using namespace std;

// @lc code=start
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int n = nums.size();

        if (n == 1) { return nums[0]==target ? 0 : -1; }
        if (n == 2) {
            return nums[0]==target ? 0 : ((nums[1]==target) ? 1 : -1);
        }
        if (nums[0] < nums[n-1]) { // No rotation
             return binarySearch(nums, 0, n-1, target);
        }

        int l = 0;
        int r = n - 1;
        int inflection = 0;

        while (l <= r) {
            if (nums[l] < nums[r]) {
                inflection = nums[l]<nums[inflection] ? l : inflection;
                break;
            }
            int m = l + ((r-l)/ 2);
            inflection = nums[inflection]<nums[m] ? inflection : m;
            if (nums[l] <= nums[m]) { l = m + 1; }
            else { r = m - 1; }
        }

        if (target < nums.back()) { // Search from inflection to nums.size()-1
            return binarySearch(nums, inflection, n-1, target);
        } else { // Search from 0 to inflection-1.
            return binarySearch(nums, 0, inflection-1, target);
        }
    }

    int binarySearch(vector<int>& nums, int l, int r, int target) {
        while (l <= r) {
            int m = l + ((r-l)/2);
            if (target < nums[m]) { r = m - 1; }
            else if (target > nums[m]) { l = m + 1; }
            else { return m; }
        }
        return -1;
    }
};
// @lc code=end

File: 33.search-in-rotated-sorted-array.py
#
# @lc app=leetcode id=33 lang=python3
#
# [33] Search in Rotated Sorted Array
#
from typing import List

# @lc code=start
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        # Run binary search to find the peak,
        # Then run binary search about the two slopes to find the target
        n = len(nums)

        if n == 1:
            return 0 if nums[0] == target else -1
        
        if nums[0] == target:
            return 0
        if nums[-1] == target:
            return len(nums) - 1

        # Finding the peak
        l, r = 0, n-1
        if nums[0] > nums[1]: # If peak is at nums[0]
            peak = 0
        elif nums[0] < nums[-1]: # If peak is at nums[-1]
            peak = n-1
        else:
            while l <= r:
                mid = l + (r - l) // 2
                
                if nums[mid] == target:
                    return mid
                if mid > 0 and nums[mid-1] == target:
                    return mid-1
                if mid < n-1 and nums[mid+1] == target:
                    return mid+1
                
                if nums[mid-1] < nums[mid] and nums[mid] > nums[mid+1]:
                    break
                else:
                    if nums[mid] > nums[0]:
                        l = mid + 1
                    else:
                        r = mid - 1
            peak = mid
        
        def binSearch(l: int, r: int):
            while l <= r:
                mid = l + (r - l) // 2
                if nums[mid] < target:
                    l = mid + 1
                elif target < nums[mid]:
                    r = mid - 1
                else:
                    return mid
            return -1

        l, r = (peak+1, n-1) if target < nums[0] else (0, peak-1)
        return binSearch(l, r)
# @lc code=end

def main():
    s = Solution()
    nums = [8,9,0,1,3,4,6]
    target = 7
    print(s.search(nums, target))

if __name__ == '__main__':
    main()
